[CASSANDRA-19759] Validate table transactional modes when preparing transactions

stonewhitener opened the pull request:
CASSANDRA-19759

stonewhitener committed:
Validate table transactional modes when preparing transactions

patch by Youki Shiraishi; reviewed by xxx for CASSANDRA-19759

dcapwell reviewed (changes requested):
Overall LGTM, but can you add tests in `org.apache.cassandra.cql3.statements.TransactionStatementTest`?

stonewhitener committed:
Make the demo transaction multi-key

stonewhitener committed:
Add tests of transactional mode validation for transaction statements

stonewhitener commented:
@dcapwell Thank you for your review comment! I added three tests for the table validation introduced in this patch and modified the existing tests affected.

dcapwell commented on test/unit/org/apache/cassandra/cql3/statements/TransactionStatementTest.java:
```diff
@@ -33,14 +33,7 @@
-import static org.apache.cassandra.cql3.statements.TransactionStatement.NO_COUNTERS_IN_TXNS_MESSAGE;
-import static org.apache.cassandra.cql3.statements.TransactionStatement.NO_TIMESTAMPS_IN_UPDATES_MESSAGE;
-import static org.apache.cassandra.cql3.statements.TransactionStatement.SELECT_REFS_NEED_COLUMN_MESSAGE;
+import static org.apache.cassandra.cql3.statements.TransactionStatement.*;
```
please revert, this goes against our style guide; we don't like `.*`

stonewhitener replied:
Done, but our `checkstyle.xml` does not include such a rule. Where can I find the evidence for your claim?

dcapwell replied:
https://cassandra.apache.org/_/development/code_style.html is what defines our style, but it overrides https://www.oracle.com/java/technologies/javase/codeconventions-fileorganization.htm

In both cases I don't see this documented... so I think we only approved moving away from `.*` in a mailing list thread and didn't update our guides?  Sorry about that =(

The only place that hits at that is `ide/idea/codeStyleSettings.xml` which makes it 100 elements before we replace with `.*`

maedhroz commented on src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java:
```diff
@@ -100,6 +100,7 @@ public class TransactionStatement implements CQLStatement.CompositeCQLStatement,
     public static final String INCOMPLETE_PRIMARY_KEY_SELECT_MESSAGE = "SELECT must specify either all primary key elements or all partition key elements and LIMIT 1. In both cases partition key elements must be always specified with equality operators; %s %s";
     public static final String NO_CONDITIONS_IN_UPDATES_MESSAGE = "Updates within transactions may not specify their own conditions; %s statement %s";
     public static final String NO_TIMESTAMPS_IN_UPDATES_MESSAGE = "Updates within transactions may not specify custom timestamps; %s statement %s";
+    public static final String TRANSACTIONS_DISABLED_ON_TABLE_MESSAGE = "Accord transactions are disabled on a table (See transactional_mode in table option); %s statement %s";
```
```suggestion
    public static final String TRANSACTIONS_DISABLED_ON_TABLE_MESSAGE = "Accord transactions are disabled on table (See transactional_mode in table options); %s statement %s";
```

stonewhitener replied:
Done! Thank you!

maedhroz reviewed (approved):
Left one minor nit, but otherwise LGTM

stonewhitener committed:
Revert imports

stonewhitener committed:
Refine error message

stonewhitener committed:
Fix tests

dcapwell commented:
merged https://github.com/apache/cassandra/commit/885f4362afdca43f0a98b03128b2b4993a6701e9
